src/main/java/com/facebook/openwifirrm/ucentral/UCentralClient.java

```diff
@@ -195,27 +199,142 @@ public boolean login() {
 		}
 
 		// Parse access token from response
-		JSONObject respBody;
+		WebTokenResult token;
 		try {
-			respBody = new JSONObject(response.getBody());
-		} catch (JSONException e) {
+			token = gson.fromJson(response.getBody(), WebTokenResult.class);
+		} catch (JsonSyntaxException e) {
 			logger.error("Login failed: Unexpected response", e);
 			logger.debug("Response body: {}", response.getBody());
 			return false;
 		}
-		if (!respBody.has("access_token")) {
+		if (
+			token == null || token.access_token == null ||
+				token.access_token.isEmpty()
+		) {
 			logger.error("Login failed: Missing access token");
-			logger.debug("Response body: {}", respBody.toString());
+			logger.debug("Response body: {}", response.toString());
 			return false;
 		}
-		this.accessToken = respBody.getString("access_token");
+		this.accessToken = token;
 		logger.info("Login successful as user: {}", username);
-		logger.debug("Access token: {}", accessToken);
+		logger.debug("Access token: {}", accessToken.access_token);
+		logger.debug("Refresh token: {}", accessToken.refresh_token);
 
 		// Load system endpoints
 		return loadSystemEndpoints();
 	}
 
+	/**
+	 * when use public endpoints, refresh the access token if it's expired.
+	 */
+	public synchronized void refreshAccessToken() {
```

Comment:
While this `synchronized` *works*, this is not a good location because it will very frequently trigger locking for no reason. You only want to lock when necessary, i.e. refreshing/re-login.

Here's one possible/simple way to only lock when needed:
```java
private WebTokenResult getAccessToken() {
	// ...
	if (isRefreshTokenExpired()) {
		synchronized (this) {			
			if (isRefreshTokenExpired()) {
				// inner logic here
			}
		}
	} else if (isAccessTokenExpired()) {
		synchronized (this) {
			if (isAccessTokenExpired()) {
				// inner logic here
			}
		}
	}
}
```
In the "normal case", no locking is done. When the token is actually expired, then all threads will block while the first thread obtains a new token, then other threads unblock and see that the token is no longer expired. Do take care to handle error cases (if first thread fails and token is now null).